platforms/core-runtime/native/src/main/java/org/gradle/platform/internal/DefaultBuildPlatform.java

```diff
@@ -0,0 +1,150 @@
+            }
+        });
+    }
+
+    public DefaultBuildPlatform(final Architecture architecture, final OperatingSystem operatingSystem) {
+        this.architecture = Suppliers.memoize(new Supplier<Architecture>() {
+            @Override
+            public Architecture get() {
+                return architecture;
+            }
+        });
+        this.operatingSystem = Suppliers.memoize(new Supplier<OperatingSystem>() {
+            @Override
+            public OperatingSystem get() {
+                return operatingSystem;
+            }
+        });
+    }
+
+    @Override
+    public Architecture getArchitecture() {
+        return architecture.get();
+    }
+
+    @Override
+    public OperatingSystem getOperatingSystem() {
+        return operatingSystem.get();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        DefaultBuildPlatform that = (DefaultBuildPlatform) o;
+        return Objects.equals(getArchitecture(), that.getArchitecture()) && Objects.equals(getOperatingSystem(), that.getOperatingSystem());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getArchitecture(), getOperatingSystem());
+    }
+
+    private static Architecture getArchitecture(SystemInfo systemInfo) {
+        SystemInfo.Architecture architecture = systemInfo.getArchitecture();
+        switch (architecture) {
+            case i386:
+                return Architecture.X86;
+            case amd64:
+                return Architecture.X86_64;
+            case aarch64:
+                return Architecture.AARCH64;
+        }
+        throw new GradleException("Unhandled system architecture: " + architecture);
+    }
+
+    public static OperatingSystem getOperatingSystem(org.gradle.internal.os.OperatingSystem operatingSystem) {
+        if (org.gradle.internal.os.OperatingSystem.LINUX == operatingSystem) {
+            return OperatingSystem.LINUX;
+        } else if (org.gradle.internal.os.OperatingSystem.UNIX == operatingSystem) {
+            return OperatingSystem.UNIX;
+        } else if (org.gradle.internal.os.OperatingSystem.WINDOWS == operatingSystem) {
+            return OperatingSystem.WINDOWS;
+        } else if (org.gradle.internal.os.OperatingSystem.MAC_OS == operatingSystem) {
+            return OperatingSystem.MAC_OS;
+        } else if (org.gradle.internal.os.OperatingSystem.SOLARIS == operatingSystem) {
+            return OperatingSystem.SOLARIS;
+        } else if (org.gradle.internal.os.OperatingSystem.FREE_BSD == operatingSystem) {
+            return OperatingSystem.FREE_BSD;
+        } else {
+            throw new GradleException("Unhandled operating system: " + operatingSystem.getName());
+        }
+    }
+
+    private void writeObject(ObjectOutputStream out) throws IOException {
```

Comment:
💭   We should prefer CC codecs for our own types over Java serialization, when there is no other need.

Here is an example of a CC codec that writes a bean-like thing:
https://github.com/gradle/gradle/blob/0a1023c144b8edfb413da5696dad61c004c110e1/platforms/core-configuration/dependency-management-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/dm/transform/TransformedArtifactCodec.kt#L41

The pattern is that we create a separate subproject with `-serialization-codecs`. I guess, to follow the pattern strictly, you’d need to create `platform-jvm-serialization-codecs`.

Since your thing is pure data, your codec would probably be stateless. Then, it can be Kotlin’s `object` instead of a class

Reply:
Addressed on https://github.com/gradle/gradle/issues/32030